File: scripts/arm.py
#! /usr/bin/env python3

import numpy as np

import sys
import time
import threading

import conn_utilities
from kortex_api.autogen.client_stubs.BaseClientRpc import BaseClient
from kortex_api.autogen.client_stubs.BaseCyclicClientRpc import BaseCyclicClient
from kortex_api.autogen.messages import Base_pb2


class Gen3Lite:
    def __init__(self):
        self.TIMEOUT_DURATION = 20
    
    
    def init(self):
        # Connection
        args = conn_utilities.parseConnectionArguments()
        self.device_conn = conn_utilities.DeviceConnection.createTcpConnection(args)
        self.device_conn.__enter__()
        self.router = self.device_conn.router
        
        # Base Client
        self.base = BaseClient(self.router)
        self.base_cyclic = BaseCyclicClient(self.router)
        
        # Serve
        base_servo_mode = Base_pb2.ServoingModeInformation()
        base_servo_mode.servoing_mode = Base_pb2.SINGLE_LEVEL_SERVOING
        self.base.SetServoingMode(base_servo_mode)
    
    
    def close(self):
        self.device_conn.__exit__(None, None, None)
    
    
    def check_for_end_or_abort(self, e):
        def check(notification, e = e):
            print("NOTIFICATION EVENT : " + \
                Base_pb2.ActionEvent.Name(notification.action_event))
            if notification.action_event == Base_pb2.ACTION_END \
            or notification.action_event == Base_pb2.ACTION_ABORT:
                e.set()
        return check

    
    def wait_for_notification(self):
        e = threading.Event()
        notification_handle = self.base.OnNotificationActionTopic(
            self.check_for_end_or_abort(e),
            Base_pb2.NotificationOptions()
        )
        finished = e.wait(self.TIMEOUT_DURATION)
        self.base.Unsubscribe(notification_handle)
        return finished
    
    
    def move_home(self):
        action_type = Base_pb2.RequestedActionType()
        action_type.action_type = Base_pb2.REACH_JOINT_ANGLES
        action_list = self.base.ReadAllActions(action_type)
        action_handle = None
        for action in action_list.action_list:
            if action.name == "Home":
                action_handle = action.handle
        if action_handle == None:
            print("RESULT : FAILURE. Home action not found.")
            return False
        self.base.ExecuteActionFromReference(action_handle)
        
        finished = self.wait_for_notification()
        print("RESULT : {}.".format("SUCCESS" if finished else "FAILURE"))
        return finished
    
    
    def set_gripper_position(self, position):
        gripper_command = Base_pb2.GripperCommand()
        gripper_command.mode = Base_pb2.GRIPPER_POSITION
        finger = gripper_command.gripper.finger.add()
        finger.value = position
        
        self.base.SendGripperCommand(gripper_command)
        time.sleep(2.5) # TODO
    
    
    def get_feedback(self):
        feedback = self.base_cyclic.RefreshFeedback()
        return feedback
    
    
    def get_current_cartesian_pose(self):
        feedback = self.get_feedback()
        current_pose = np.array([
            feedback.base.tool_pose_x,
            feedback.base.tool_pose_y,
            feedback.base.tool_pose_z,
            feedback.base.tool_pose_theta_x,
            feedback.base.tool_pose_theta_y,
            feedback.base.tool_pose_theta_z
        ])
        return current_pose
    
    
    def get_current_cartesian_velocity(self):
        feedback = self.get_feedback()
        current_velocity = np.array([
            feedback.base.tool_twist_linear_x,
            feedback.base.tool_twist_linear_y,
            feedback.base.tool_twist_linear_z,
            feedback.base.tool_twist_angular_x,
            feedback.base.tool_twist_angular_y,
            feedback.base.tool_twist_angular_z
        ])
        return current_velocity

    
    def get_current_angular_pose(self):
        feedback = self.get_feedback()
        current_thetas = np.array([feedback.actuators[idx].position for idx in range(6)])
        return current_thetas
    
    
    def get_current_angular_velocity(self):
        feedback = self.get_feedback()
        current_thetas_dot = np.array([feedback.actuators[idx].velocity for idx in range(6)])
        return current_thetas_dot
    
    
    def get_current_external_wrench(self):
        feedback = self.get_feedback()
        current_external_wrench = np.array([
            feedback.base.tool_external_wrench_force_x,
            feedback.base.tool_external_wrench_force_y,
            feedback.base.tool_external_wrench_force_z,
            feedback.base.tool_external_wrench_torque_x,
            feedback.base.tool_external_wrench_torque_y,
            feedback.base.tool_external_wrench_torque_z
        ])
        return current_external_wrench
    
    
    def set_cartesian_pose(self, cartesian_pose):
        action = Base_pb2.Action()

        target_cartesian_pose = action.reach_pose.target_pose
        target_cartesian_pose.x = cartesian_pose[0]
        target_cartesian_pose.y = cartesian_pose[1]
        target_cartesian_pose.z = cartesian_pose[2]
        target_cartesian_pose.theta_x = cartesian_pose[3]
        target_cartesian_pose.theta_y = cartesian_pose[4]
        target_cartesian_pose.theta_z = cartesian_pose[5]

        self.base.ExecuteAction(action)

        finished = self.wait_for_notification()
        print("RESULT : {}.".format("SUCCESS" if finished else "FAILURE"))
        return finished

    
    def set_angular_pose(self, thetas):
        action = Base_pb2.Action()
        
        for idx in range(6):
            joint_angle = action.reach_joint_angles.joint_angles.joint_angles.add()
            joint_angle.joint_identifier = idx
            joint_angle.value = thetas[idx]

        self.base.ExecuteAction(action)

        finished = self.wait_for_notification()
        print("RESULT : {}.".format("SUCCESS" if finished else "FAILURE"))
        return finished
    
    
    def stop_motioning(self):
        self.base.Stop()
    
    
    def set_joint_speeds(self, thetas_d):
        joint_speeds = Base_pb2.JointSpeeds()
        
        for idx in range(6):
            joint_speed = joint_speeds.joint_speeds.add()
            joint_speed.joint_identifier = idx
            joint_speed.value = thetas_d[idx]
            joint_speed.duration = 0
        
        self.base.SendJointSpeedsCommand(joint_speeds)
    
    
    def set_twist(self, twists, reference_frame = Base_pb2.CARTESIAN_REFERENCE_FRAME_BASE):
        command = Base_pb2.TwistCommand()
        command.reference_frame = reference_frame
        command.duration = 0
        
        twist = command.twist
        twist.linear_x = twists[0]
        twist.linear_y = twists[1]
        twist.linear_z = twists[2]
        twist.angular_x = twists[3]
        twist.angular_y = twists[4]
        twist.angular_z = twists[5]
        
        self.base.SendTwistCommand(command)

File: cpp/utils/arm.cpp
#include "arm.h"

#define CONNECTION_PORT 10000

constexpr auto TIMEOUT_DURATION = std::chrono::seconds{20};

Gen3Lite::Gen3Lite(CommanderArgs args) {
    // Create API objects
    auto error_callback = [](k_api::KError err){cout << "Callback error." << err.toString();};
    transport = std::make_shared<k_api::TransportClientTcp>();
    router = std::make_shared<k_api::RouterClient>(transport.get(), error_callback);
    transport->connect(args.ip_address, CONNECTION_PORT);

    // Set session data connection information
    auto session_info = k_api::Session::CreateSessionInfo();
    session_info.set_username(args.username);
    session_info.set_password(args.password);
    session_info.set_session_inactivity_timeout(60000);   // ms
    session_info.set_connection_inactivity_timeout(2000); // ms

    // Session manager service wrapper
    std::cout << "Creating session for communication..." << std::endl;
    session_manager = std::make_shared<k_api::SessionManager>(router.get());
    session_manager->CreateSession(session_info);
    std::cout << "Session created." << std::endl;

    // Create services
    base = std::make_shared<k_api::Base::BaseClient>(router.get());
    base_cyclic = std::make_shared<k_api::BaseCyclic::BaseCyclicClient>(router.get());

    // Start services
    auto servoingMode = k_api::Base::ServoingModeInformation();
    servoingMode.set_servoing_mode(k_api::Base::ServoingMode::SINGLE_LEVEL_SERVOING);
    base->SetServoingMode(servoingMode);
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
}

Gen3Lite::~Gen3Lite() {
    session_manager->CloseSession();
    router->SetActivationStatus(false);
    transport->disconnect();
}

std::function<void(k_api::Base::ActionNotification)> event_listener_by_promise(std::promise<k_api::Base::ActionEvent>& finish_promise) {
    return [&finish_promise] (k_api::Base::ActionNotification notification) {
        const auto action_event = notification.action_event();
        switch(action_event) {
            case k_api::Base::ActionEvent::ACTION_END:
                std::cout << "[Notification Event] ACTION_END." << std::endl;
                finish_promise.set_value(action_event);
                break;
            case k_api::Base::ActionEvent::ACTION_ABORT:
                std::cout << "[Notification Event] ACTION_ABORT." << std::endl;
                finish_promise.set_value(action_event);
                break;
            default:
                break;
        }
    };
}

bool Gen3Lite::wait_for_notification() {
    std::promise<k_api::Base::ActionEvent> finish_promise;
    auto finish_future = finish_promise.get_future();
    auto promise_notification_handle = base->OnNotificationActionTopic(
        event_listener_by_promise(finish_promise),
        k_api::Common::NotificationOptions()
    );
    const auto status = finish_future.wait_for(TIMEOUT_DURATION);
    base->Unsubscribe(promise_notification_handle);
    // const auto promise_event = finish_future.get();
    return status == std::future_status::ready;
}

k_api::BaseCyclic::Feedback Gen3Lite::GetFeedback() {
    auto feedback = base_cyclic->RefreshFeedback();
    return feedback;
}

Vector6f Gen3Lite::GetToolCartesianPose() {
    auto feedback = base_cyclic->RefreshFeedback();
    Vector6f pose;
    pose <<
        feedback.base().tool_pose_x(),
        feedback.base().tool_pose_y(),
        feedback.base().tool_pose_z(),
        feedback.base().tool_pose_theta_x(),
        feedback.base().tool_pose_theta_y(),
        feedback.base().tool_pose_theta_z();
    return pose;
}

Vector6f Gen3Lite::GetToolTwist() {
    auto feedback = base_cyclic->RefreshFeedback();
    Vector6f twist;
    twist <<
        feedback.base().tool_twist_linear_x(),
        feedback.base().tool_twist_linear_y(),
        feedback.base().tool_twist_linear_z(),
        feedback.base().tool_twist_angular_x(),
        feedback.base().tool_twist_angular_y(),
        feedback.base().tool_twist_angular_z();
    return twist;
}

bool Gen3Lite::MoveArmToHome() {
    auto action_type = k_api::Base::RequestedActionType();
    action_type.set_action_type(k_api::Base::REACH_JOINT_ANGLES);
    auto action_list = base->ReadAllActions(action_type);
    auto action_handle = k_api::Base::ActionHandle();
    action_handle.set_identifier(0);
    for (auto action : action_list.action_list()) {
        if (action.name() == "Home") {
            action_handle = action.handle();
        }
    }
    if (action_handle.identifier() == 0) {
        std::cout << "[Error] Home cannot be reached." << std::endl;
        return false;
    }

    base->ExecuteActionFromReference(action_handle);

    bool finished = wait_for_notification();
    if (finished) {
        std::cout << "[Action] Succeed." << std::endl;
    } else {
        std::cout << "[Action] Failed." << std::endl;
    }
    return finished;
}

bool Gen3Lite::SetGripperPosition(float position) {
    k_api::Base::GripperCommand gripper_command;
    gripper_command.set_mode(k_api::Base::GRIPPER_POSITION);
    auto finger = gripper_command.mutable_gripper()->add_finger();
    finger->set_finger_identifier(1);
    finger->set_value(position);

    base->SendGripperCommand(gripper_command);
    std::this_thread::sleep_for(std::chrono::milliseconds(2500));   // TODO
}

bool Gen3Lite::SetCartesianPose(Vector6f target_pose) {
    auto action = k_api::Base::Action();
    action.set_application_data("");

    auto constrained_pose = action.mutable_reach_pose();
    auto pose = constrained_pose->mutable_target_pose();
    pose->set_x(target_pose(0));
    pose->set_y(target_pose(1));
    pose->set_z(target_pose(2));
    pose->set_theta_x(target_pose(3));
    pose->set_theta_y(target_pose(4));
    pose->set_theta_z(target_pose(5));

    base->ExecuteAction(action);

    bool finished = wait_for_notification();
    if (finished) {
        std::cout << "[Action] Succeed." << std::endl;
    } else {
        std::cout << "[Action] Failed." << std::endl;
    }
    return finished;
}

void Gen3Lite::StopMotioning() {
    base->Stop();
}

void Gen3Lite::SetTwist(Vector6f target_twist, k_api::Common::CartesianReferenceFrame reference_frame) {
    auto command = k_api::Base::TwistCommand();
    command.set_reference_frame(reference_frame);
    command.set_duration(0);

    auto twist = command.mutable_twist();
    twist->set_linear_x(target_twist(0));
    twist->set_linear_y(target_twist(1));
    twist->set_linear_z(target_twist(2));
    twist->set_angular_x(target_twist(3));
    twist->set_angular_y(target_twist(4));
    twist->set_angular_z(target_twist(5));

    base->SendTwistCommand(command);
}
